### PR TITLE
Add wasmrelaxedsimd x8-packw

### PR DESCRIPTION
**Pull request by fanchenkong1**

### CODE REVIEW

File: src/x8-packw/kr-wasmdot.c.in [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        $for N in range(NR):
+          w${N} += ${KR};
+        out += ${NR*KR};
+      }
+
+      // KC remainder 1..KR-1
+      if (k != 0) {
+        assert(k >= 1 && k <= ${KR-1});
+
+        const v128_t all_one = wasm_i32x4_splat(-1);
```
Comment (fbarchard) [inline src/x8-packw/kr-wasmdot.c.in:121]: vone preferred for value of 1 or vmask for value of -1

Comment (fanchenkong1) [reply src/x8-packw/kr-wasmdot.c.in]: Done.

File: src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        w7 += 8;
+        out += 64;
+      }
+
+      // KC remainder 1..KR-1
+      if (k != 0) {
+        assert(k >= 1 && k <= 7);
+
+        const v128_t all_one = wasm_i32x4_splat(-1);
+        const v128_t mask = wasm_u64x2_shr(all_one, (8 - k) * sizeof(int8_t) * 8);
```
Comment (fbarchard) [inline src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c]: vector variables should start with 'v'.    vmask.

File: src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        w6 += 8;
+        w7 += 8;
+        out += 64;
+      }
+
+      // KC remainder 1..KR-1
+      if (k != 0) {
+        assert(k >= 1 && k <= 7);
+
+        const v128_t vmask = wasm_u64x2_shr(wasm_i32x4_splat(-1), (8 - k) * sizeof(int8_t) * 8);
```
Comment (fbarchard) [inline src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c:167]: nice... i've always meant to try this technique.
its similar to how avx512 would generate a k mask
there is also masked loads for avx512, but only for float, but byte.
this shift technique only works up to 8 bytes I think.  most cpus dont natively have a bit shift of 128 bits or more... after that it becomes a byte shuffle, but the shuffler is usually a vector, so this approach starts to fall down.

This is so much simplier than the avx code that does inserts of 4, 2 and 1 bytes.
For future 4 bit, I think a shuffle is also needed.
e.g.  read 6 bytes, but move them to the first 3 bytes, a pad, and then next 3 bytes.

**Review by fbarchard — APPROVED** [review]

**Review by fbarchard — APPROVED** [review]

**Comment by fanchenkong1** [comment]
This change is rebased on the latest commit. PTAL, thanks!

File: src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        v128_t v0_01 = wasm_v128_load(w0);
+        v128_t v1_01 = wasm_v128_load(w1);
+        v128_t v2_01 = wasm_v128_load(w2);
+        v128_t v3_01 = wasm_v128_load(w3);
+        v128_t v4_01 = wasm_v128_load(w4);
+        v128_t v5_01 = wasm_v128_load(w5);
+        v128_t v6_01 = wasm_v128_load(w6);
+        v128_t v7_01 = wasm_v128_load(w7);
+
+        v128_t v01_0 = wasm_i64x2_shuffle(v0_01, v1_01, 0, 2);
```
Comment (fbarchard) [inline src/qs8-packw/gen/qs8-packw-x8c8-gemm-goi-wasmrelaxedsimd.c:93]: i expect revec can't handle this.
for avx I was thinking of doing 4 columns, but its complicated.  the first 2 columns are as you have it.  the next 2 columns i think can be done with insertf128

**Review by fbarchard — APPROVED** [review]
overall awesome!  i like you 2 column loop and your KC remainder masking.
We should do that on native vnni as well
I think your code will still work without vnni, but you might want to think of a way to do the sum for regular wasmsimd.  The hard way isnt all that hard
convert to int32 and sum

your NR remainder should really be a copy/paste of the entire main loop with only 2 changes at the top
1. NC bias copy/zero
2. NC clip the pointers
I think the main NR loop could be adapted to handle these, cutting the amount of code in half, and in your case, make the NC remainder do the 2 column KR loop, speeding it up.
But it would rule out some optimizations on the main loop, so I think leave it for now.
The main loop, on native, is doing a poor job on pointers.  It might help to literally only have the 1 pointer - w0, and use stride to index the rows.

**Pull request merged**